### PR TITLE
IGAPP-389: Refactor location permission handling

### LINKED ISSUE #389: No permission permission dialog if denied forever
**Describe the bug**
If you deny location permission forever, then the app is no longer helping you to enable it.

**To Reproduce**
1. Disable location permission
2. Click on "Show nearby"

**Expected behavior**
Should show a dialog which leads you to the setting

**Screenshots**

DB030A50-B55D-4838-9E41-3CCCDF08B895

https://user-images.githubusercontent.com/905221/137977715-5ef7487a-523c-4564-8a21-eed8bb4dd548.mp4

**Smartphone (please complete the following information):**
 - Device: iPhone X

**Additional context**

### LINKED ISSUE #393: Location shown as denied after permitting it
**Describe the bug**
When going through the intro slides, permitting location access causes the text to indicate missing permission anyways.

**To Reproduce**
https://user-images.githubusercontent.com/13063245/137990802-9907e422-b42a-49b2-ac68-b73094dfed8a.mp4

**Expected behavior**
The text on the disabled button should indicate that permissions were granted

**Smartphone (please complete the following information):**
 - Device: Realme irgendwas
 - OS:  Android 11
 - Version 2.0.2 (Test version)

### PR DESCRIPTION
**Pull request by maxammann**
fixes #389 

fixes #393 

TODO: Write descripction about PR

### CODE REVIEW

File: frontend/lib/intro_slides/location_request_button.dart [hunk 1]
```diff
@@ -10,55 +11,49 @@ class LocationRequestButton extends StatefulWidget {
...
         return const ElevatedButton(
           onPressed: null,
           child: Text("Standort ist bereits freigegeben."),
         );
         break;
-      case _LocationStatus.denied:
+      case LocationPermission.deniedForever:
         return const ElevatedButton(
           onPressed: null,
           child: Text("Standort ist nicht freigegeben."),
```
Comment (Schedulaar) [inline frontend/lib/intro_slides/location_request_button.dart:59]: I would expect to be able to click the button even though something is wrong with the permission (i.e. lead to user to the system menu etc).
Maybe you can add that.
Here the button is somewhat strangely used as a status message.
The button should imo only be disabled, if the location permission is already given correctly.

Comment (Schedulaar) [inline frontend/lib/intro_slides/location_request_button.dart:59]: This is also displayed when location services are disabled. We should have a different text in this case.

**Review by Schedulaar — CHANGES_REQUESTED** [review]

File: frontend/lib/intro_slides/location_request_button.dart (hunk 1 above)
Comment (maxammann) [reply frontend/lib/intro_slides/location_request_button.dart:59]: Fixed

File: frontend/lib/location/determine_position.dart [hunk 2]
```diff
@@ -1,97 +1,95 @@
...
+      final result = await showDialog(
+          context: context,
           builder: (context) => const LocationServiceDialog());
-      if (result == true) {
+      if (result) {
         await Geolocator.openLocationSettings();
       }
     }
-    throw PositionNotAvailableException('Location service was disabled.');
+    return LocationPermission.deniedForever;
```
[the remark this answers is not in the document]
Comment (maxammann) [reply frontend/lib/location/determine_position.dart:61]: I added a custom enum

File: frontend/lib/location/determine_position.dart [hunk 3]
```diff
@@ -1,97 +1,95 @@
...
 /// Ensures all preconditions needed to determine the current position.
 /// If needed, location permissions are requested.
 ///
-/// When the location services are not enabled or permissions
-/// are denied the `Future` will return with an `PositionNotAvailableException`.
-/// Else it will complete without a value.
-Future<void> requestPermissionToDeterminePosition(
-    {BuildContext userInteractContext}) async {
+/// When the location services are not enabled then it will return
+/// LocationPermission.deniedForever
```
[the remark this answers is not in the document]
Comment (maxammann) [reply frontend/lib/location/determine_position.dart:47]: I added a custom enum

File: frontend/lib/intro_slides/location_request_button.dart (hunk 1 above)
Comment (maxammann) [reply frontend/lib/intro_slides/location_request_button.dart:59]: Fixed

**Comment by maxammann** [comment]
Fixes will be done in https://github.com/ehrenamtskarte/ehrenamtskarte/pull/398

**Review by Schedulaar — APPROVED** [review]

**Pull request merged**